### ISSUE TITLE
`in` Operator Narrowing is Too Limited

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#50891]
[Playground Link](https://www.typescriptlang.org/play?ts=4.9.0-dev.20220921#code/JYOwLgpgTgZghgYwgAgEoHEBCyDeBYAKGWOSggBMAuZEAVwFsAjaAbkJOQHMyIRq6mrdiUYAbWhH4NmUNgQC+hQjFogEYYAHsQyAM4QwAYU2jNUABQA3OOMnJVAaxCaA7iACUuYcQTbdYZGBdDGwAXm8OZAB6KMDdZDgdTUYAKwh1CI5rW2QAMlzkMABPAAcITRhkbIlkULrkACJktPUGvNzMkgaycjbQKpsa-MLS8srqiAA6Htr6hoEZNvzO4gbuCF4+nQn2kbKKgdtJ9d5Z0MaF6CWOokjVsQktw6GC4v3xwamHlDrz+ekrnIIsBKuYgiFPPhbndfCB-KROIxqCFas8IHIOIoFEA)

```ts
interface RGB {
    red: number;
    green: number;
    blue: number;
}

function setColor(value: unknown) {
    const isRGB =
        // is an object
        value && typeof value === "object" &&
        // 'red' exists and is a 'number'
        "red" in value && typeof value.red === "number" &&
        // 'green' exists and is a 'number'
        "green" in value && typeof value.green === "number" &&
        // 'blue' exists and is a 'number'
        "blue" in value && typeof value.blue === "number";

    if (isRGB) {
        const rgb: RGB = value;
    }
}
```

**Expected**: No errors
**Actual** Errors.

```
Type 'object & Record<"red", unknown> & Record<"green", unknown> & Record<"blue", unknown>' is not assignable to type 'RGB'.
  Types of property 'red' are incompatible.
    Type 'unknown' is not assignable to type 'number'.
```

**Comment by ahejlsberg** [comment]
This is basically a duplicate of #42384 (which in turn is a duplicate of a much older issue for the same scenario, but I can't locate that right now). You'd see the same problem if you started out with a `{ red: unknown, green: unknown, blue: unknown }` object and applied just the `typeof` checks.

**Comment by ahejlsberg** [comment]
It's interesting, though... We could consider having a `typeof value.red === "number"` narrow the type of `value` by intersecting with `Record<"red", number>`, and we could do similar narrowing for other types of tests:

```ts
type Point = { x: number, y: number };

declare function doWithPoint(p: Point): void;

function test(p: Partial<Point>) {
    if (p.x && p.y) {
        doWithPoint(p); // Narrowed to Partial<Point> & Record<"x", number> & Record<"y", number>
    }
}
```

One concern would be the noise created by all the intersections, but it does seem like a workable idea.
